katdal/sensordata.py

```diff
@@ -55,50 +55,110 @@
 
 
 class SensorData(object):
-    """Raw (uncached) sensor data in record array form.
+    """Raw (uninterpolated) sensor data wrapper.
 
-    This is a convenient container for uncached sensor data which resembles
-    a record array with fields ('timestamp', 'value', 'status'). This is also
-    the typical format of HDF5 datasets used to store sensor data. Its main
-    advantage is that it exposes the sensor data dtype (from the 'value' field)
-    as a top-level attribute, making it compatible with NumPy arrays and
-    :class:`CategoricalData` objects when used together in a sensor cache. It
-    also exposes the sensor name, if available.
+    This is a convenient wrapper for uninterpolated sensor data which resembles
+    a structured array with fields 'timestamp', 'value' and optionally 'status'.
+
+    Its main advantage is that it exposes the sensor data dtype (from the
+    'value' field) as a top-level attribute, making it compatible with NumPy
+    arrays and :class:`CategoricalData` objects when used together in a sensor
+    cache. It also exposes the sensor name, if available.
+
+    The idea is that the raw sensor data is not initially cached in this object,
+    making it a light-weight wrapper focussing on sensor metadata. All data
+    access should be via __getitem__ of the appropriate field.
 
     Parameters
     ----------
-    data : recarray-like, with fields ('timestamp', 'value', 'status')
-        Uncached sensor data as record array or equivalent (such as a
-        :class:`h5py.Dataset`)
-    name : string or None, optional
-        Sensor name (assumed to be data.name by default, if it exists)
+    name : string
+        Sensor name
+    dtype : :class:`numpy.dtype` object
+        Sensor data type as NumPy dtype (beware that string lengths may be wrong)
```

Comment:
It *is* an attribute on the base class that is set by the derived classes, and this is its only docstring. I wanted to highlight that the dtype detection is a bit iffy when working with variable-length strings as classes typically only check the dtype of the first or last element to be efficient, and the resulting string length might be too short for other strings in the dataset.

For this very reason I am thinking of ditching ```dtype``` as a required parameter and only allowing it on ```RecordSensorData```. The sensor extractor funnels everything through this class on the way to ```CategoricalData``` or ndarrays (see ```remove_duplicates_and_failures``` function).